emissary-core/src/primitives/mapping.rs

```diff
@@ -18,102 +18,127 @@
 
 use crate::primitives::Str;
 
-use hashbrown::HashMap;
+use bytes::{BufMut, BytesMut};
+use hashbrown::{
+    hash_map::{IntoIter, Iter},
+    HashMap,
+};
 use nom::{
     number::complete::{be_u16, be_u8},
     IResult,
 };
 
-use alloc::{vec, vec::Vec};
-use core::fmt;
+use alloc::vec::Vec;
+use core::{
+    fmt::{self, Debug},
+    num::NonZeroUsize,
+};
 
 /// Key-value mapping
-#[derive(Debug, PartialEq, Eq)]
-pub struct Mapping {
-    /// Key
-    key: Str,
-
-    /// Value.
-    value: Str,
-}
+#[derive(Debug, PartialEq, Eq, Clone, Default)]
+pub struct Mapping(HashMap<Str, Str>);
 
 impl fmt::Display for Mapping {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        write!(f, "{}={}", self.key, self.value)
+        self.0.fmt(f)
     }
 }
 
 impl Mapping {
-    /// Create new [`Mapping`].
-    pub fn new(key: Str, value: Str) -> Self {
-        Self { key, value }
-    }
-
     /// Serialize [`Mapping`] into a byte vector.
-    pub fn serialize(self) -> Vec<u8> {
-        let key = self.key.serialize();
-        let value = self.value.serialize();
-
-        // key length + value length + length field + `=` + `;`
-        let size = key.len() + value.len() + 2;
-        let mut out = vec![0u8; size];
-
-        out[..key.len()].copy_from_slice(&key);
-        out[key.len()] = b'=';
-        out[1 + key.len()..1 + key.len() + value.len()].copy_from_slice(&value);
-        out[1 + key.len() + value.len()] = b';';
+    pub fn serialize(&self) -> Vec<u8> {
+        // Allocate at least two bytes for the size prefix
+        let mut out = BytesMut::with_capacity(2);
+        let mut data = out.split_off(2);
+
+        let mut entries: Vec<_> = self.0.iter().collect();
+
+        // Our mapping implementation does not support duplicate keys, so we do not need to preserve
+        // order
+        entries.sort_unstable_by(|a, b| a.0.cmp(b.0));
+        for (key, value) in entries {
+            let key = key.serialize();
+            let value = value.serialize();
+            data.reserve(key.len() + value.len() + 2);
+            data.extend(key);
+            data.put_u8(b'=');
+            data.extend(value);
+            data.put_u8(b';');
+        }
+        //TODO: validate that this is less than u16::MAX?
+        out.put_u16(data.len() as u16);
+        out.unsplit(data);
 
-        out
+        //TODO: Work with `Bytes` instead of `Vec<u8>` to avoid copying data?
+        out.freeze().to_vec()
```

Comment:
Returning `Bytes`/`BytesMut` is preferred. Most of the primitives were written before `bytes` was added which is why some `serialize()` methods return `Vec<u8>` and others return `Bytes`/`BytesMut`